merkle/src/tree_impl.rs

```diff
@@ -0,0 +1,192 @@
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Creates a new instance of [MinaMerkleTree] with estimated capacity of leaves
+    pub fn with_capacity(capacity: usize) -> Self {
+        let protential_depth = calculate_depth(capacity);
+        let protential_node_count = calculate_node_count(protential_depth);
+        Self {
+            leafs: Vec::with_capacity(capacity),
+            nodes: Vec::with_capacity(protential_node_count),
+            ..Default::default()
+        }
+    }
+
+    /// Clears cached hashes of all ancester nodes of the give leaf
+    /// because the values become invaid once the leaf is updated
+    fn clear_dirty_hashes(&mut self, leaf_index: usize) {
+        let mut parent = leaf_index;
+        while parent > 0 {
+            parent = calculate_parent_index(parent);
+            if self.nodes[parent].is_some() {
+                self.nodes[parent] = None;
+            } else {
+                break;
+            }
+        }
+    }
+
+    /// Calucates hash of a node if it's not available in the node cache
+    /// either apply hash algorithm if it's a leaf node
+    /// or apply merge algorithm if it's a non-leaf node
+    /// update the cache once calculated
+    fn calculate_hash_if_needed(
+        &mut self,
+        index: usize,
+    ) -> Option<(THash, MerkleTreeNodeMetadata<DEGREE>)> {
+        if index < self.nodes.len() {
+            if let Some(hash) = &self.nodes[index] {
+                Some((hash.clone(), MerkleTreeNodeMetadata::new(index)))
+            } else {
+                let left = index * 2 + 1;
+                let right = index * 2 + 2;
+                let left_hash = self.calculate_hash_if_needed(left);
+                let right_hash = self.calculate_hash_if_needed(right);
+                let hash = TMerger::merge([&left_hash, &right_hash]);
+                self.nodes[index] = hash.clone();
+                hash.map(|hash| (hash, MerkleTreeNodeMetadata::new(index)))
+            }
+        } else {
+            let leaf_index = index - self.nodes.len();
+            if leaf_index < self.leafs.len() {
+                let (hash, _) = &self.leafs[leaf_index];
+                Some((hash.clone(), MerkleTreeNodeMetadata::new(index)))
+            } else {
+                None
+            }
+        }
+    }
+}
+
+impl<TItem, THash, THasher, TMerger> MerkleTree<DEGREE>
+    for MinaMerkleTree<TItem, THash, THasher, TMerger>
+where
+    THasher: MerkleHasher<DEGREE, Item = TItem, Hash = THash>,
+    TMerger: MerkleMerger<DEGREE, Hash = THash>,
+    THash: Clone,
+{
+    type Item = TItem;
+    type Hash = THash;
+
+    fn depth(&self) -> u32 {
+        self.depth
+    }
+
+    fn count(&self) -> usize {
+        self.leafs.len()
+    }
+
+    fn root(&mut self) -> Option<Self::Hash> {
+        self.calculate_hash_if_needed(0).map(|(hash, _)| hash)
+    }
+
+    fn add_batch(&mut self, items: Vec<Self::Item>) {
```

Comment:
Please add rustdoc and normal comments for all of this, even internal things.

Reply:
This is trait implementation, the generated rust doc will automatically inherit the doc from the Trait definition, any rust doc here will override the inherited doc.

To verify this, run `cargo doc --no-deps --release --all-features` and check generated doc under target/docs/mina_merkle